src/Microsoft.AspNet.OData.Shared/Formatter/Deserialization/ODataResourceSetDeserializer.cs

```diff
@@ -147,9 +167,143 @@ public virtual IEnumerable ReadResourceSet(ODataResourceSetWrapper resourceSet,
                     Error.Format(SRResources.TypeCannotBeDeserialized, elementType.FullName()));
             }
 
-            foreach (ODataResourceWrapper resourceWrapper in resourceSet.Resources)
+            //Ideally we don't need to special case ResourceSetType.ResourceSet, since the code that handles a deltaresourcesetwrapper will also handle a ResourceSetWrapper, 
+            //but it may be more efficient for the common case.
+
+            if (resourceSet.ResourceSetType == ResourceSetType.ResourceSet)
+            {
+                foreach (ODataResourceWrapper resourceWrapper in resourceSet.Items)
+                {
+                    yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                }
+            }
+            else
+            {
+                Type clrType = EdmLibHelpers.GetClrType(elementType, readContext.Model);
+
+                if (readContext.IsChangedObjectCollection)
+                {
+                    readContext.ResourceType = typeof(Delta<>).MakeGenericType(clrType);
+                }
+
+                foreach (ODataItemBase odataItemBase in resourceSet.Items)
+                {
+                    ODataResourceWrapper resourceWrapper = odataItemBase as ODataResourceWrapper;
+
+                    if(resourceWrapper != null)
+                    {
+                        ODataDeletedResource deletedResource = resourceWrapper.ResourceBase as ODataDeletedResource;
+
+                        if (deletedResource != null)
+                        {
+                            yield return DeSerializeDeletedEntity(deletedResource, readContext, clrType);
+                        }
+                        else
+                        {
+                            yield return deserializer.ReadInline(resourceWrapper, elementType, readContext);
+                        }
+                    }
+                    else
+                    {
+                        ODataDeltaLinkWrapper deltaLinkWrapper = odataItemBase as ODataDeltaLinkWrapper;
+
+                        Contract.Assert(deltaLinkWrapper != null, "ODataDeltaLinkWrapper should not be null");
+
+                        yield return DeserializeDeltaLink(deltaLinkWrapper, resourceSet.ResourceSetBase.TypeName, readContext);
+                    }                    
+                }
+            }
+        }
+
+
+        private static IEdmDeltaLinkBase DeserializeDeltaLink(ODataDeltaLinkWrapper deltaLinkWrapper, string typeName, ODataDeserializerContext readContext)
+        {
+            ODataDeltaLinkBase deltalink = deltaLinkWrapper.DeltaLink;
+
+            Contract.Assert(deltalink != null, "ODataDeltaLink should not be null");
+
+            IEdmModel model = readContext.Model;
+                
+            if (model == null)
+            {
+                throw Error.Argument("readContext", SRResources.ModelMissingFromReadContext);
+            }
```

Comment:
should this be an assert?  is it our bug if readContext has no model, or something the developer did wrong?